enhancements/results-forwarding.md

```diff
@@ -0,0 +1,379 @@
+  
+  - It's non-trivial to do disaster recovery: An object snapshot contains
+    CRD metadata which is hard to replicate (e.g. object UUID) and is irrelevant
+    for compliance checks. Full etcd restore is a tedious and complicated process [1].
+
+  - Values are limited to a size of 1.5Mb which makes evidence storage
+    tricky and expensive. e.g. we currently use one PVC per `ComplianceScan`
+    for raw results which may be considered evidence.
+
+  - The current CRDs may be very verbose, which may result in big objects, which
+    may incur a penalty in performance and scalability for Kubernetes. [2]
+
+While projects such as StackRox [3] have successfully integrated Compliance Operator
+via CRDs, having a more lightweight approach 
+
+By sending results and evidences to a central store, we'll be able to:
+
+* View Compliance results and suggested fixes for multiple clusters in one place.
+
+* Address the aforementioned limitations with Etcd by choosing an alternative
+  store. e.g. a relational database or a cloud-managed database.
+
+The Compliance Operator also stores raw results (currently ARF) in a PVC by default.
+While this is useful, it has proven problematic in multiple deployments given that
+not everybody has access to persistent storage in their clusters.
+
+ARF files contain a lot of information about the state of the system where the
+compliance scan was effectuated. We could take the view that this is "evidence"
+gathered for the system in check. With this in mind, we can generalize and
+expand the concept to allow for forwarding this evidence to an alternative
+object store, e.g. S3.
+
+In the future, we could expand the "evidence" gathered, e.g. for Kubernetes
+compliance scans we could simply compress the gathered objects the
+`api-resource-collector` downloads and send that too.
+
+### Goals
+
+* Provide a generic way to implement compliance result forwarding for Compliance Operator.
+  This should be flexible enough to accommodate the current implementation (CRDs and PVCs),
+  while allowing for custom forwarder implementations.
+
+* Implement and define a stable API for forwarder implementations.
+
+* Implement result/remediation CRDs and ARF PVC as reference implementations of the forwarding
+  model.
+
+### Non-Goals
+
+* Deprecate the current CRD/PVC approach: There is still a use-case for this.
+
+* Expand the "evidence" that's currently gathered by the Compliance Operator. This is
+  needed but should be done as part of a separate initiative.
+
+## Proposal
+
+In short, the proposal is to have Compliance Operator *always* forward Compliance results (suggested
+remediations being part of this), as well as evidence (e.g. ARF results). This will require
+changing the **aggregator** to always forward via a GRPC to a configured implementation. The
+**resultserver** then becomes subject to the evidence forwarding provider and will be renamed
+**evidence-persistor**.
```

Comment:
Can you expand on the relationship between the provider and the forwarder and how are ARF results going to be processed?
Right now the flow is this:

- scans run
- when the scan pod finishes, the scan pod itself creates the configMap with the ARF results and uploads the ARF file into the resultserver
- when ALL scan pod finish, the aggregator kicks in and creates the CRs

As far as I understand, the aggregator would now instead of creating the CRs forward data into the grpc service, this is the forwarder. I like this part.

I'm confused about the resultcollector though. Would we keep the resultcollector as a place that would upload the ARFs to a different location based on config or forward the ARFs through grpc? Note that a quick search shows that grpc message is limited to 4MB.

I guess we could even make the resultserver image just configurable (unless we already do it?) and people could then just swap their own resultserver. The original would just sit there, other implementations might have a resultserver that forwards the results to S3 etc. But this doesn't solve the problem of having a PVC for the resultserver..

Reply:
yeah, GRPC doesn't work well for big binaries; I think we need to keep the current implementation very similar to how it is now, and make the resultcollector hold the results in memory and start forwarding them ASAP.